### ISSUE TITLE
Crashes when clicking back button to exit on Android

### ISSUE THREAD

**Issue by DVassilev** [airbnb/react-native-maps#414]
Hi, I use MapView as the main screen of the app and when I try exiting using the back button it crashes with the following error
```java.lang.NullPointerException: Attempt to invoke interface method 'void maps.af.y.q()' on a null object reference
	at maps.af.Q.d(Unknown Source)```
Anyone know how I can fix it?
Thanks

**Comment by DavidYKay** [comment]
My users are getting this error left and right as well. I haven't gotten to the bottom of it yet, but would love to figure this out.

**Comment by DVassilev** [comment]
@DavidYKay I can't seem to get to the bottom of it either. Putting logs on this here too.

```
Fatal Exception: java.lang.NullPointerException: Attempt to invoke interface method 'void maps.ad.z.q()' on a null object reference
       at maps.ad.R.d(Unknown Source)
       at wc.onTransact(:com.google.android.gms.DynamiteModulesB:87)
       at android.os.Binder.transact(Binder.java:387)
       at com.google.android.gms.maps.internal.IMapViewDelegate$zza$zza.onDestroy(Unknown Source)
       at com.google.android.gms.maps.MapView$zza.onDestroy(Unknown Source)
       at com.google.android.gms.dynamic.zza.onDestroy(Unknown Source)
       at com.google.android.gms.maps.MapView.onDestroy(Unknown Source)
       at com.airbnb.android.react.maps.AirMapView.doDestroy(AirMapView.java:289)
       at com.airbnb.android.react.maps.AirMapManager.onDropViewInstance(AirMapManager.java:70)
       at com.airbnb.android.react.maps.AirMapManager.onDropViewInstance(AirMapManager.java:27)
       at com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:514)
       at com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:523)
       at com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:523)
       at com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:523)
       at com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:523)
       at com.facebook.react.uimanager.NativeViewHierarchyManager.manageChildren(NativeViewHierarchyManager.java:414)
       at com.facebook.react.uimanager.UIViewOperationQueue$ManageChildrenOperation.execute(UIViewOperationQueue.java:175)
       at com.facebook.react.uimanager.UIViewOperationQueue$2.run(UIViewOperationQueue.java:782)
       at com.facebook.react.uimanager.UIViewOperationQueue.flushPendingBatches(UIViewOperationQueue.java:829)
       at com.facebook.react.uimanager.UIViewOperationQueue.access$1500(UIViewOperationQueue.java:44)
       at com.facebook.react.uimanager.UIViewOperationQueue$3.run(UIViewOperationQueue.java:807)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:234)
       at android.app.ActivityThread.main(ActivityThread.java:5526)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

**Comment by DavidYKay** [comment]
Based on what I'm reading, it's about passing in the Activity Context vs the
Application Context. Which is a huge bummer, because the new approach that RN
takes is to ONLY use the Application Context.  

  

Not sure what to do about this. I ended up moving my app to 100% native code.
:|

![](https://link.nylas.com/open/7lgnhhtx4kky6uuo5cmrw3py1/local-
796ffa42-fe43?r=cmVwbHkrMDAwMjlkMTcyODJkZDdhMzliYjgyYmJhODI4YjJhYmVhZDEwOTY3N2E5ZWY3YTZlOTJjZjAwMDAwMDAxMTNiNDQyNjg5MmExNjljZTA5ZjM5YTQ0QHJlcGx5LmdpdGh1Yi5jb20=)

On Jul 30 2016, at 3:24 am, DVassilev &lt;notifications@github.com&gt; wrote:  

>
[@DavidYKay](https://github.com/DavidYKay&r=cmVwbHkrMDAwMjlkMTcyODJkZDdhMzliYjgyYmJhODI4YjJhYmVhZDEwOTY3N2E5ZWY3YTZlOTJjZjAwMDAwMDAxMTNiNDQyNjg5MmExNjljZTA5ZjM5YTQ0QHJlcGx5LmdpdGh1Yi5jb20=)
I can't seem to get to the bottom of it either. Putting logs on this here too.

>  
>  
>     Fatal Exception: java.lang.NullPointerException: Attempt to invoke
interface method 'void maps.ad.z.q()' on a null object reference

>            at maps.ad.R.d(Unknown Source)

>            at wc.onTransact(:com.google.android.gms.DynamiteModulesB:87)

>            at android.os.Binder.transact(Binder.java:387)

>            at
com.google.android.gms.maps.internal.IMapViewDelegate$zza$zza.onDestroy(Unknown
Source)

>            at com.google.android.gms.maps.MapView$zza.onDestroy(Unknown
Source)

>            at com.google.android.gms.dynamic.zza.onDestroy(Unknown Source)

>            at com.google.android.gms.maps.MapView.onDestroy(Unknown Source)

>            at
com.airbnb.android.react.maps.AirMapView.doDestroy(AirMapView.java:289)

>            at
com.airbnb.android.react.maps.AirMapManager.onDropViewInstance(AirMapManager.java:70)

>            at
com.airbnb.android.react.maps.AirMapManager.onDropViewInstance(AirMapManager.java:27)

>            at
com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:514)

>            at
com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:523)

>            at
com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:523)

>            at
com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:523)

>            at
com.facebook.react.uimanager.NativeViewHierarchyManager.dropView(NativeViewHierarchyManager.java:523)

>            at
com.facebook.react.uimanager.NativeViewHierarchyManager.manageChildren(NativeViewHierarchyManager.java:414)

>            at
com.facebook.react.uimanager.UIViewOperationQueue$ManageChildrenOperation.execute(UIViewOperationQueue.java:175)

>            at
com.facebook.react.uimanager.UIViewOperationQueue$2.run(UIViewOperationQueue.java:782)

>            at
com.facebook.react.uimanager.UIViewOperationQueue.flushPendingBatches(UIViewOperationQueue.java:829)

>            at
com.facebook.react.uimanager.UIViewOperationQueue.access$1500(UIViewOperationQueue.java:44)

>            at
com.facebook.react.uimanager.UIViewOperationQueue$3.run(UIViewOperationQueue.java:807)

>            at android.os.Handler.handleCallback(Handler.java:739)

>            at android.os.Handler.dispatchMessage(Handler.java:95)

>            at android.os.Looper.loop(Looper.java:234)

>            at android.app.ActivityThread.main(ActivityThread.java:5526)

>            at java.lang.reflect.Method.invoke(Method.java)

>            at
com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)

>            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)

>  
>

> —  
> You are receiving this because you were mentioned.  
> Reply to this email directly, [view it on
GitHub](https://github.com/lelandrichardson/react-native-
maps/issues/414#issuecomment-236357700&r=cmVwbHkrMDAwMjlkMTcyODJkZDdhMzliYjgyYmJhODI4YjJhYmVhZDEwOTY3N2E5ZWY3YTZlOTJjZjAwMDAwMDAxMTNiNDQyNjg5MmExNjljZTA5ZjM5YTQ0QHJlcGx5LmdpdGh1Yi5jb20=),
or [mute the thread](https://github.com/notifications/unsubscribe-
auth/AAKdF0ONnkK1AcOLpROOPdwCu7GY4KPtks5qayZogaJpZM4JSZXV&r=cmVwbHkrMDAwMjlkMTcyODJkZDdhMzliYjgyYmJhODI4YjJhYmVhZDEwOTY3N2E5ZWY3YTZlOTJjZjAwMDAwMDAxMTNiNDQyNjg5MmExNjljZTA5ZjM5YTQ0QHJlcGx5LmdpdGh1Yi5jb20=).![](https://github.com/notifications/beacon/AAKdF3bYvv8pav_T1OmegSj4OlVOG7qMks5qayZogaJpZM4JSZXV.gif)

**Comment by alvinthen** [comment]
Temporary fixed by catching the exception, this will only suppress the crash with unkown side effects

In `node_modules/react-native-maps/android/lib/src/main/java/com/airbnb/android/react/maps/AirMapView.java` - `doDestroy()`

```java
try {
    onDestroy();
} catch (Exception e) {
    // Suppress crash when back button is pressed, unknown side effects
    Log.e("AirMapView", e.getMessage());
}
```

**Comment by BaptisteT** [comment]
Is there any plan for a long term fix ?

**Comment by mauron85** [comment]
Didn't spent much time reading source, but would help adding/replacing LifecycleEventListener with its method onHostDestroy? I did it in my own module.

https://github.com/mauron85/react-native-background-geolocation/blob/master/android/lib/src/main/java/com/marianhello/react/BackgroundGeolocationModule.java

**Comment by alvinthen** [comment]
PR #449 is now passing Activity to MapInitializer, but it doesn't seem to fix this, can anyone verify?

**Comment by alexHlebnikov** [comment]
Still have this error. Is there any real way to fix it?

**Comment by ciAnd7** [comment]
Hi,
This patch of `AirMapView.java` 0.8.2 works for me:

```
--- /Temp/0.8.2/AirMapView.java	2016-09-20 13:27:10.000000000 +0300
+++ node_modules/react-native-maps/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java	2016-09-20 10:43:01.000000000 +0300
@@ -86,6 +86,7 @@
     private LifecycleEventListener lifecycleListener;
     private OnLayoutChangeListener onLayoutChangeListener;
     private boolean paused = false;
+    private boolean destroyed = false;
     private ThemedReactContext context;
 
     final EventDispatcher eventDispatcher;
@@ -288,7 +289,10 @@
         if (!paused) {
             onPause();
         }
-        onDestroy();
+        if (!destroyed) {
+            destroyed = true;
+            onDestroy();
+        }
     }
 
     public void setRegion(ReadableMap region) {

```

**Comment by cymen** [comment]
I can consistently reproduce this on a Nexus 5 with Android 6.0.1 installed with the July 1, 2016 Android security patches. That would be build MMB30K.

I then installed the next month security patches (August) and the crash is gone. The MMB30K build can be downloaded here: https://developers.google.com/android/nexus/images

**Comment by cymen** [comment]
I forgot to say I tried applying the `onDestroy()` work round above and that patch did not stop the crash.

**Comment by alexHlebnikov** [comment]
Same here - workaround doesn't helped.

**Comment by kkgelu** [comment]
I can't believe the issue is from Jul and still no progress made. I end up exiting the app from another view:

```
      BackAndroid.addEventListener('hardwareBackPress', () => {
       if (currentState === MAP_VIEW) {
         //  set to another state to render without the map or
         //  navigate to another scene without map view
         return true;
       }
       return false;
      });

```

**Comment by GeneralRamy** [comment]
I handle this issue as @mikelambert mention in [com/airbnb/android/react/maps/AirMapView.java](https://github.com/airbnb/react-native-maps/blob/master/android/src/main/java/com/airbnb/android/react/maps/AirMapView.java)
i think app trying to do onDestory  to paused view without moving to onPause which out of android lift cycle 

  `if (!paused) {
            onPause();
            onDestroy();
   }`

**Comment by spikebrehm** [comment]
Should be fixed by #694, which i just merged.
